File: tests/coretests/Stocks/PriceBarTests.cs
using System;
using core.fs.Shared.Adapters.Stocks;
using coretests.testdata;
using Microsoft.FSharp.Core;
using Xunit;

namespace coretests.Stocks;

public class PriceBarTests
{
    private readonly PriceBar _bar = TestDataGenerator.PriceBars(TestDataGenerator.NET).First;
    
    [Fact]
    public void ClosingRange_Works() => Assert.Equal(0.75m, _bar.ClosingRange(), 2);
    
    [Fact]
    public void PercentDifferenceFromHigh_Works() => Assert.Equal(0.02m, _bar.PercentDifferenceFromHigh(_bar.Close), 2);
    
    [Fact]
    public void PercentDifferenceFromLow_Works() => Assert.Equal(-0.07m, _bar.PercentDifferenceFromLow(_bar.Close), 2);
    
    [Fact]
    public void DateStr_Works() => Assert.Equal("2020-11-30", _bar.DateStr);
    
    [Fact]
    public void TrueRange_Correct() => Assert.Equal(6.67m, _bar.TrueRange(FSharpOption<PriceBar>.None), 2);
}

public class PriceBarsTests
{
    private readonly PriceBars _bars = TestDataGenerator.PriceBars(TestDataGenerator.NET);
    
    [Fact]
    public void Length() => Assert.Equal(505, _bars.Length);

    [Fact]
    public void AllButLast()
    {
        Assert.Equal(504, _bars.AllButLast().Length);
        Assert.Equal("2020-11-30", _bars.AllButLast().First.DateStr);
        Assert.Equal("2022-11-29", _bars.AllButLast().Last.DateStr);
    }
    
    [Fact]
    public void Last() => Assert.Equal("2022-11-30", _bars.Last.DateStr);

    [Fact]
    public void LatestOrAll_WithSmallerRange()
    {
        var subset = _bars.LatestOrAll(numberOfBars: 20);
        
        Assert.Equal(20, subset.Length);
        Assert.Equal("2022-11-30", subset.Last.DateStr);
        Assert.Equal("2022-11-02", subset.First.DateStr);
    }

    [Fact]
    public void ClosingPrices()
    {
        Assert.Equal(505, _bars.ClosingPrices().Length);
        Assert.Equal(75.08m, _bars.ClosingPrices()[0]);
        Assert.Equal(49.14m, _bars.ClosingPrices()[504]);
    }

    [Fact]
    public void Volumes()
    {
        Assert.Equal(505, _bars.Volumes().Length);
        Assert.Equal(17157618m, _bars.Volumes()[0]);
        Assert.Equal(6933219m, _bars.Volumes()[504]);
    }
    
    [Fact]
    public void FindByDate_ReturnsNone_WhenNotFound()
    {
        var result = _bars.TryFindByDate(DateTimeOffset.Parse("2020-01-01"));
        Assert.True(FSharpOption<Tuple<PriceBar,int>>.get_IsNone(result));
    }
    
    [Fact]
    public void FindByDate_ReturnsSome_WhenFound()
    {
        var result = _bars.TryFindByDate(DateTimeOffset.Parse("2020-11-30"));
        Assert.True(FSharpOption<Tuple<PriceBar,int>>.get_IsSome(result));
        Assert.Equal("2020-11-30", result.Value.Item1.DateStr);
    }
}

File: tests/coretests.fs/Stocks/PriceBarTests.fs
module coretests.fs.Stocks.PriceBarTests

open Xunit
open FsUnit
open testutils

let bars = TestDataGenerator.PriceBars(TestDataGenerator.NET)
let bar = bars.First

[<Fact>]
let ``Closing range works``() =
    Assert.Equal(0.75m, bar.ClosingRange(), 2);

let ``Percent difference from high works``() =
    Assert.Equal(0.02m, bar.PercentDifferenceFromHigh(bar.Close), 2)
    
[<Fact>]
let ``Percent difference from low works``() =
    Assert.Equal(-0.07m, bar.PercentDifferenceFromLow(bar.Close), 2)

[<Fact>]
let ``DateStr works``() =
    Assert.Equal("2020-11-30", bar.DateStr)

[<Fact>]
let ``True range works``() =
    Assert.Equal(6.67m, bar.TrueRange(None), 2)

[<Fact>]
let ``Length works``() =
    bars.Length |> should equal 505
    
[<Fact>]
let ``AllButLast works``() =
    bars.AllButLast().Length |> should equal 504
    bars.AllButLast().First.DateStr |> should equal "2020-11-30"
    bars.AllButLast().Last.DateStr |> should equal "2022-11-29"

[<Fact>]
let ``First works``() =
    bars.First.DateStr |> should equal "2020-11-30"

[<Fact>]
let ``Last works``() =
    bars.Last.DateStr |> should equal "2022-11-30"

[<Fact>]
let ``LatestOrAll with smaller range works``() =
    let subset = bars.LatestOrAll(numberOfBars = 20)
    subset.Length |> should equal 20
    subset.Last.DateStr |> should equal "2022-11-30"
    subset.First.DateStr |> should equal "2022-11-02"

[<Fact>]
let ``Closing Prices accurate``() =
    bars.ClosingPrices().Length |> should equal 505
    bars.ClosingPrices()[0] |> should equal 75.08m
    bars.ClosingPrices()[504] |> should equal 49.14m
    
[<Fact>]
let ``Volumes accurate`` () =
    bars.Volumes().Length |> should equal 505
    bars.Volumes()[0] |> should equal 17157618m
    bars.Volumes()[504] |> should equal 6933219m
    
[<Fact>]
let ``Find by date returns none when not found``() =
    let result = bars.TryFindByDate(System.DateTimeOffset.Parse("2020-01-01"))
    result |> should equal None
    
[<Fact>]
let ``Find by date returns some when found``() =
    let result = bars.TryFindByDate(System.DateTimeOffset.Parse("2020-11-30"))
    result |> should equal (Some(bar, 0))